File: src/core/cache/redis_manager.py
import json
from typing import Optional, Any
from redis.asyncio import Redis, ConnectionPool
from datetime import datetime, timezone

from src.config import get_settings


class RedisManager:
    def __init__(self, redis_url: str, default_ttl: int = 300):
        self.pool = ConnectionPool.from_url(redis_url, decode_responses=True)
        self.client = Redis(connection_pool=self.pool)
        self.default_ttl = default_ttl

    async def cache_certificate_status(
        self,
        certificate_id: str,
        is_valid: bool,
        expires_at: datetime,
        data: Optional[dict] = None,
        ttl: Optional[int] = None,
    ):
        """Cache certificate validity status"""
        ttl = ttl or self.default_ttl
        await self.client.setex(f"cert:status:{certificate_id}", ttl, json.dumps(data))

    async def get_certificate_status(self, certificate_id: str) -> Optional[dict]:
        """Get cached certificate status"""
        data = await self.client.get(f"cert:status:{certificate_id}")
        if data:
            return json.loads(data)
        return None

    async def invalidate_certificate(self, certificate_id: str):
        """Invalidate all cache entries for a certificate"""
        await self.client.delete(f"cert:status:{certificate_id}")

    async def check_rate_limit(
        self, key: str, max_requests: int, window_seconds: int
    ) -> bool:
        """Sliding window rate limiter"""
        now = datetime.now(timezone.utc).timestamp()
        window_key = f"rate_limit:{key}"

        # Remove old entries
        await self.client.zremrangebyscore(window_key, 0, now - window_seconds)

        # Count current requests
        current_count = await self.client.zcard(window_key)

        if current_count >= max_requests:
            return False

        # Add current request
        await self.client.zadd(window_key, {str(now): now})
        await self.client.expire(window_key, window_seconds)
        return True

    async def close(self):
        await self.client.close()
        await self.pool.disconnect()


def get_redis() -> RedisManager:
    return RedisManager(
        get_settings().redis_url, get_settings().redis_cache_ttl_seconds
    )

File: app/Services/RedisManager.php
<?php

namespace App\Services;

use DateTimeInterface;
use Illuminate\Support\Facades\Redis;

class RedisManager
{
    protected int $defaultTtl;

    /**
     * Pass configuration defaults via constructor.
     */
    public function __construct(int $defaultTtl = 300)
    {
        $this->defaultTtl = $defaultTtl;
    }

    /**
     * Cache certificate validity status.
     */
    public function cacheCertificateStatus(
        string $certificateId, 
        bool $isValid, 
        DateTimeInterface $expiresAt, 
        ?array $data = null, 
        ?int $ttl = null
    ): void {
        $ttl = $ttl ?? $this->defaultTtl;
        
        // Ensure data exists, appending validation context if needed
        $payload = $data ?? [
            'is_valid' => $isValid,
            'expires_at' => $expiresAt->format(DateTimeInterface::ATOM)
        ];

        // setex sets value with TTL in seconds
        Redis::setex("cert:status:{$certificateId}", $ttl, json_encode($payload));
    }

    /**
     * Get cached certificate status.
     */
    public function getCertificateStatus(string $certificateId): ?array
    {
        $data = Redis::get("cert:status:{$certificateId}");
        
        if ($data) {
            return json_decode($data, true);
        }

        return null;
    }

    /**
     * Invalidate all cache entries for a certificate.
     */
    public function invalidateCertificate(string $certificateId): void
    {
        Redis::del("cert:status:{$certificateId}");
    }

    /**
     * Precise Sliding Window Rate Limiter.
     * Uses Redis Sorted Sets (ZSET) to block API bursts.
     */
    public function checkRateLimit(string $key, int $maxRequests, int $windowSeconds): bool
    {
        $now = microtime(true); // High resolution timestamp float
        $windowKey = "rate_limit:{$key}";
        
        // 1. Remove old logged request scores outside the current time window sliding range
        Redis::zremrangebyscore($windowKey, 0, $now - $windowSeconds);

        // 2. Count current active records in the set
        $currentCount = Redis::zcard($windowKey);

        if ($currentCount >= $maxRequests) {
            return false;
        }

        // 3. Log the current request timestamp (using a unique member string value)
        // syntax: zadd(key, score, member)
        Redis::zadd($windowKey, $now, (string)$now);
        
        // 4. Ensure the key auto-destructs after the window expires
        Redis::expire($windowKey, $windowSeconds);

        return true;
    }
}